src/dmd/dtoh.d

```diff
@@ -0,0 +1,1724 @@
+            ParseTimeVisitor    = Identifier.idPool("ParseTimeVisitor");
+        }
+
+    }
+
+    private bool isIdentifierClass(ASTCodegen.ClassDeclaration cd)
+    {
+        return (cd.ident == DMDClass.ID &&
+                cd.parent !is null &&
+                cd.parent.ident == DMDModule.identifier &&
+                cd.parent.parent && cd.parent.parent.ident == DMDModule.dmd &&
+                !cd.parent.parent.parent);
+    }
+
+    private bool isVisitorClass(ASTCodegen.ClassDeclaration cd)
+    {
+        for (auto cdb = cd; cdb; cdb = cdb.baseClass)
+        {
+            if (cdb.ident == DMDClass.Visitor ||
+                cdb.ident == DMDClass.ParseTimeVisitor)
+            return true;
+        }
+        return false;
+    }
+
+    private bool isIgnoredModule(ASTCodegen.Module m)
+    {
+        //if (!m || !m.parent)
+        if (!m)
+            return true;
+
+        // Ignore dmd.root
+        if (m.parent && m.parent.ident == DMDModule.root &&
+            m.parent.parent && m.parent.parent.ident == DMDModule.dmd &&
+            !m.parent.parent.parent)
+        {
+            return true;
+        }
+
+        // Ignore dmd.visitor and derivatives
+        if ((m.ident == DMDModule.visitor ||
+             m.ident == DMDModule.parsetimevisitor ||
+             m.ident == DMDModule.permissivevisitor ||
+             m.ident == DMDModule.strictvisitor ||
+             m.ident == DMDModule.transitivevisitor) &&
+             m.parent && m.parent.ident == DMDModule.dmd &&
+             !m.parent.parent)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    private bool isFrontendModule(ASTCodegen.Module m)
+    {
+        if (!m || !m.parent)
+            return false;
+
+        // Ignore dmd.root
+        if (m.parent.ident == DMDModule.root &&
+            m.parent.parent && m.parent.parent.ident == DMDModule.dmd &&
+            !m.parent.parent.parent)
+        {
+            return false;
+        }
+
+        // Ignore dmd.visitor and derivatives
+        if ((m.ident == DMDModule.visitor ||
+             m.ident == DMDModule.parsetimevisitor ||
+             m.ident == DMDModule.permissivevisitor ||
+             m.ident == DMDModule.strictvisitor ||
+             m.ident == DMDModule.transitivevisitor) &&
+             m.parent && m.parent.ident == DMDModule.dmd &&
+             !m.parent.parent)
+        {
+            return false;
+        }
+        return ((m.parent.ident == DMDModule.dmd && !m.parent.parent) ||
+                (m.parent.parent.ident == DMDModule.dmd && !m.parent.parent.parent));
+    }
+//}
+
+void genCppHdrFiles(ref Modules ms)
+{
+    DMDType._init();
+    //version(BUILD_COMPILER)
+    //{
+        DMDModule._init();
+        DMDClass._init();
+    //}
+
+    OutBuffer buf;
+    buf.writestring("#pragma once\n");
+    buf.writeByte('\n');
+    buf.printf("// Automatically generated by dmd -HC\n");
```

Comment:
I'm wondering whether we should include the DMD version. 
On one hand it's informative, on the other hand it would need to needless diff in the event a generated header is checked in and its generation is part of the process (it doesn't make sense to me, but I'm sure someone will do it).
Same issue would arise with a timestamp. However I can see a value in knowing a bit more about how the header was generated.

Last but not least, having the full command line might be helpful too. Things like `-version` and `-debug` will affect the output, won't they ?

Reply:
Indeed, that could be useful to track down errors but also annoying for diffs (and maybe reproducible builds). Maybe this behavior should be customizable via an optional argument to `-HC`. But i think this could be added later.

> Last but not least, having the full command line might be helpful too. Things like `-version` and `-debug` will affect the output, won't they ?

I think so.

Reply:
At the very least, use `__VENDOR__`, as it's not only dmd that would use this option.